staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker_test.go

```diff
@@ -0,0 +1,283 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package feature
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apiserver/pkg/storage"
+)
+
+type mockEndpointVersion struct {
+	Endpoint string
+	Version  string
+	Error    error
+}
+
+// MockEtcdClient is a mock implementation of the EtcdClientInterface interface.
+type MockEtcdClient struct {
+	EndpointVersion []mockEndpointVersion
+}
+
+func (m MockEtcdClient) getEndpoints() []string {
+	var endpoints []string
+	for _, ev := range m.EndpointVersion {
+		endpoints = append(endpoints, ev.Endpoint)
+	}
+	return endpoints
+}
+
+func (m MockEtcdClient) getVersion(endpoint string) (string, error) {
+	for _, ev := range m.EndpointVersion {
+		if ev.Endpoint == endpoint {
+			return ev.Version, ev.Error
+		}
+	}
+	// Never should happen, unless tests having a problem.
+	return "", fmt.Errorf("No version found")
+}
+
+func (m *MockEtcdClient) Endpoints() []string {
+	return m.getEndpoints()
+}
+
+// Status returns a mock status response.
+func (m *MockEtcdClient) Status(ctx context.Context, endpoint string) (*clientv3.StatusResponse, error) {
+	version, err := m.getVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	// Return a mock status response
+	return &clientv3.StatusResponse{
+		Version: version,
+	}, nil
+}
+
+func TestSupports(t *testing.T) {
+	tests := []struct {
+		testName       string
+		featureName    string
+		expectedResult bool
+		expectedError  error
+	}{
+		{
+			testName:      "Error with unknown feature",
+			featureName:   "some unknown feature",
+			expectedError: fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", "some unknown feature"),
+		},
+		{
+			testName:      "Error with empty feature",
+			featureName:   "",
+			expectedError: fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", ""),
+		},
+		{
+			testName:       "No error but disabled by default",
+			featureName:    storage.RequestWatchProgress,
+			expectedResult: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.testName, func(t *testing.T) {
+			var testFeatureSupportChecker FeatureSupportChecker = newDefaultFeatureSupportChecker()
+
+			supported, err := testFeatureSupportChecker.Supports(tt.featureName)
+
+			assert.Equal(t, tt.expectedResult, supported)
+			assert.Equal(t, tt.expectedError, err)
+		})
+	}
+}
+
+func TestSupportsRequestWatchProgress(t *testing.T) {
+	type testCase struct {
+		endpointsVersion []mockEndpointVersion
+		expectedResult   bool
+		expectedError    error
+	}
+	tests := []struct {
+		testName string
+		rounds   []testCase
+	}{
+		{
+			testName: "Default disabled",
+			rounds:   []testCase{{expectedResult: false}},
+		},
+		{
+			testName: "Enabled supported versions bound",
+			rounds: []testCase{
```

Comment:
Not sure if the following formatting is servicing readability.

Reply:
Agreed but it's easier to test complex scenarios this way. I can happily changed it to previous one if you think that's better.

Reply:
I meant code format, not contents.

Reply:
Ops, goformat is making it that way.